### PR TITLE
drivers: atmel-rstc: add support for reset controller

### PR DESCRIPTION
**Pull request by clementleger**
This reset controller will be used by PSCI to reset the SoC.

Signed-off-by: Clément Léger <clement.leger@bootlin.com>

### CODE REVIEW

File: core/drivers/atmel_rstc.c [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Microchip
+ */
+
+#include <compiler.h>
+#include <drivers/atmel_rstc.h>
+#include <io.h>
+#include <kernel/boot.h>
```
Comment (jforissier) [inline core/drivers/atmel_rstc.c:9]: Not needed?

File: core/drivers/atmel_rstc.c [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#include <kernel/dt.h>
+#include <libfdt.h>
```
Comment (jforissier) [inline core/drivers/atmel_rstc.c:11]: Not needed?

File: core/drivers/atmel_rstc.c [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#include <stdbool.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
+#include <trace.h>
```
Comment (jforissier) [inline core/drivers/atmel_rstc.c:15]: Not needed?

**Review by jforissier — COMMENTED** [review]
Minor comments below, then:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by clementleger** [comment]
New version:
 - Cleanup in includes

File: core/include/drivers/atmel_rstc.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+bool atmel_rstc_available(void);
+
+void __noreturn atmel_rstc_reset(void);
+#else
+static inline bool atmel_rstc_available(void)
+{
+	return false;
+}
+
+static inline void __noreturn atmel_rstc_reset(void) {}
```
[the remark this answers is not in the document]
Comment (clementleger) [reply core/include/drivers/atmel_rstc.h]: Yes.

File: core/drivers/atmel_rstc.c [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+void __noreturn atmel_rstc_reset(void)
+{
+	uint32_t val = AT91_RSTC_CR_KEY | AT91_RSTC_CR_PROCRST |
+		       AT91_RSTC_CR_PERRST;
+
+	io_write32(rstc_base + AT91_RSTC_CR, val);
+
+	while (1)
```
[the remark this answers is not in the document]
Comment (clementleger) [reply core/drivers/atmel_rstc.c]: Ok, I will use `while (true)`

File: core/drivers/atmel_rstc.c [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+#include <drivers/atmel_rstc.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
+
+#define AT91_RSTC_CR		0x0
+#define   AT91_RSTC_CR_KEY	(0xA5 << 24)
```
[the remark this answers is not in the document]
Comment (clementleger) [reply core/drivers/atmel_rstc.c]: Ack

**Comment by clementleger** [comment]
New version:
- Removed `__noreturn` from inline version of reset
- Added comment for reset
- Used `while(true)`
- use `SHIFT_U32` for define

File: core/drivers/atmel_rstc.c [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+#include <drivers/atmel_rstc.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
+
+#define AT91_RSTC_CR		0x0
+#define   AT91_RSTC_CR_KEY	SHIFT_U32(0xA5, 24)
```
Comment (jenswi-linaro) [inline core/drivers/atmel_rstc.c:13]: Please remove the extra spaces between define and the name, here and below.

File: core/include/drivers/atmel_rstc.h [hunk 8]
```diff
@@ -0,0 +1,24 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Microchip
+ */
+#ifndef ATMEL_RSTC_H
```
Comment (jenswi-linaro) [inline core/include/drivers/atmel_rstc.h:5]: Please add a `__DRIVERS_` prefix. The `__` part is optional, but I prefer using it.

File: core/drivers/atmel_rstc.c [hunk 9]
```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, Microchip
+ */
+
+#include <drivers/atmel_rstc.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
```
Comment (jenswi-linaro) [inline core/drivers/atmel_rstc.c:10]: Nit: add `types_ext.h` for `vaddr_t`.

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Pull request merged**